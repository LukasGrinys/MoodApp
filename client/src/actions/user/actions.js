export const actions = {
    createUser : {
        request : 'CREATE_USER_REQUEST',
        receive : 'CREATE_USER_RECEIVE',
        reset: 'CREATE_USER_RESET'
    },
    loginUser : {
        request : 'LOGIN_USER_REQUEST',
        receive : 'LOGIN_USER_RECEIVE'
    },
    clearUserData : 'CLEAR_USER',
    userAuth: 'USER_AUTH',
    editUser : {
        request : 'EDIT_USER_REQUEST',
        receive : 'EDIT_USER_RECEIVE',
        reset : 'EDIT_USER_RESET'
    },
    changePassword : {
        request : 'CHANGE_PASSWORD_REQUEST',
        receive : 'CHANGE_PASSWORD_RECEIVE',
        reset : 'CHANGE_PASSWORD_RESET'
    }
}