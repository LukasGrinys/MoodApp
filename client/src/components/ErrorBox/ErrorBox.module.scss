@import '../../styles/config/variables';

.container {
    animation-name: fadeIn;
    animation-duration: 0.5s;
    background: $color-error-background;
    border: 1px solid $color-error;
    border-radius: 2px;
    color: $color-error;
    font-size: 12px;
    margin-top: 8px;
    opacity: 1;
    padding: 8px 12px;
    text-align: center;
}

@keyframes fadeIn {
    0% { opacity: 0 };
    100% { opacity: 1};
}