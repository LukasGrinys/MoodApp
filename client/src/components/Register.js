import React from 'react';
import FormWrapper from './../widgets/FormWrapper';
import RegisterForm from './registerForm';
const Register = () => {
    return (
        <div>
            <FormWrapper>
                <RegisterForm/>
            </FormWrapper>
        </div>
    );
};

export default Register;