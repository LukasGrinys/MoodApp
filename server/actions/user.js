const { db } = require('../database/firebase');
const { COLLECTION_NAMES } = require('../database/constants');

const {
    hashPassword,
    doesUserExist,
    findUserByEmail,
    comparePassword,
    generateToken,
    findUserByToken,
    deleteToken
} = require('../util/helpers');

const { validator } = require('../util/validator');
const userSchema = require('../schemas/userSchema');

const registerNewUser = async (body) => {
    if (!body.password) {
        return {error: "No password was provided"}
    };

    const { error, data } = validator({
        email: body.email,
        firstName: body.firstName,
        lastName: body.lastName,
        password: body.password
    }, userSchema);

    if (error || !data) return { error : error || 'Values were not valid' }

    try {
        const { firstName, lastName, email, password } = data;

        const userExists = await doesUserExist(email);
        if (userExists) return  { error : "User already exists" }

        const hash = await hashPassword(password);

        if (hash) {
            await db.collection(COLLECTION_NAMES.users).doc().set({
                email,
                firstName,
                lastName,
                password: hash,
                notifications: true,
                token: ''
            });

            return { success : true }
        }
    } catch (error) {
        return { error }
    }
}

const loginUser = async (email, password) => {
    try {
        const user = await findUserByEmail(email);
        if (!user || user.error || !user.data) return {
            error: user.error ? user.error : 'User not found'
        }
        
        const isMatch = await comparePassword(password, user.data.password);
        if (!isMatch) return { error : 'Wrong password'}

        const { error, token } = generateToken(user.data.id);
        if (error || !token) return {error: 'Error: could not generate new token'};

        if (token) {
            await db.collection(COLLECTION_NAMES.users).doc(user.data.id).update({token});

            return {
                data: {
                    token,
                    id: user.data.id,
                    email: user.data.email
                }
            }
        }
    } catch (error) {
        return { error }
    }
}

const logOutUser = async (token) => {
    const user = await findUserByToken(token);

    if (user.error || !user.data || !user.data.id ) return { error : error || 'Error : Could not find user'}

    const userId = user.data.id;

    const { error } = await deleteToken(userId);

    if (error) return { error };
    
    return { success : true }
}

const editAccount = async (id, details) => {
    const { error, data } = validator({
        firstName: details.firstName,
        lastName: details.lastName
    }, userSchema);

    if (error || !data ) return { error : error || 'Error : Validation error occured'}

    try {
        const {
            firstName,
            lastName
        } = data;

        await db.collection(COLLECTION_NAMES.users).doc(id).update({
            firstName,
            lastName
        });

        return { success : true }
    } catch (error) {
        return { error }
    }  
}

const getUser = async (id) => {
    try {
        const userRef = await db.collection(COLLECTION_NAMES.users).doc(id).get();

        if (!userRef.exists) {
            return {error : "User not found"}
        }

        const data = userRef.data();
        delete data.password;

        return {
            data : {
                id,
                ...data
            }
        }
    } catch (error) {
        return {error}
    }
}

const changePassword = async (id, oldPassword, newPassword) => {
    try {
        const user = await db.collection(COLLECTION_NAMES.users).doc(id).get();
        const userData = user.data();
        
        const isMatch = await comparePassword(oldPassword, userData.password);

        if (!isMatch) return { error : 'Wrong password'}

        const { error, data } = validator({
            password: newPassword
        }, userSchema);
    
        if (error || !data) return { error : error || 'Error : Validation error occured'}
        
        const { password } = data;

        const hash = await hashPassword(password);

        await db.collection(COLLECTION_NAMES.users).doc(id).update({
            password: hash
        })

        return { success : true }
    } catch(error) {
        return { error }
    }
}

const deleteUser = async (id, password) => {
    try {
        const user = await db.collection(COLLECTION_NAMES.users).doc(id).get();
        const userData = user.data();

        const isMatch = await comparePassword(password, userData.password);
        if (!isMatch) return { error : 'Wrong password'}

        await db.collection(COLLECTION_NAMES.users).doc(id).delete();

        return { success : true }
    } catch (error) {
        console.log({error})
        return { error }
    }
}

module.exports = { 
    registerNewUser,
    loginUser,
    logOutUser,
    editAccount,
    getUser,
    changePassword,
    deleteUser
};
