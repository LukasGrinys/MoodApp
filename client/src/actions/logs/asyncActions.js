import axios from 'axios';
import { actions } from './actions';

export const canUserLog = async (currentDate, currentTiming, user) => {
    const body = {
        userId: user,
        date: currentDate,
        timing: currentTiming
    };

    return async dispatch => {
        const { data } = await axios.post('/api/canLog', body);

        dispatch({type: actions.canLog.request})
        
        if (data && data.data) {
            dispatch({
                type: actions.canLog.receive,
                payload: data.data.canLog
            })

            return;
        }

        dispatch({
            type: actions.canLog.receive,
            payload: false
        })
    }
}

export const getLastLogs = async ({
    userId,
    limit = 9
}) => {
    const url = `/api/getLogs?id=${userId}&limit=${limit}`;

    return async dispatch => {
        dispatch({type: actions.getLastLogs.request});

        try {
            const { data } = await axios.get(url);
            const { success, error } = data;
    
            if (error) {
                dispatch({
                    type: actions.getLastLogs.receive,
                    payload: {
                        logs: [],
                        error: error,
                        isFetching: false
                    }
                });
    
                return;
            }
    
            if (success && data.data) {
                dispatch({
                    type: actions.getLastLogs.receive,
                    payload: {
                        logs: data.data,
                        error: null,
                        isFetching: false
                    }
                });
    
                return;
            }
        } catch (error) {
            dispatch({
                type: actions.getLastLogs.receive,
                payload: {
                    logs: [],
                    error: error,
                    isFetching: false
                }
            });
        }
    }
}

export const clearLastLogs = () => dispatch => dispatch({type: actions.getLastLogs.reset});

export const getLog = async logId => {
    return async dispatch => {
        const endpoint = `/api/getLog?id=${logId}`;

        dispatch({type: actions.getLog.request});

        try {
            const response = await axios.get(endpoint);
            
            const { error, data } = response.data;

            if (error) {
                dispatch({
                    type: actions.getLog.receive,
                    payload: {
                        isFetching : false,
                        error,
                        logData : {}
                    }
                });

                return;
            }
            
            dispatch({
                type: actions.getLog.receive,
                payload: {
                    isFetching : false,
                    error : false,
                    logData : data
                }
            });
        } catch (error) {
            dispatch({
                type: actions.getLog.receive,
                payload: {
                    isFetching : false,
                    error,
                    logData : {}
                }
            });
        }
    }
}

export const clearLog = () => dispatch => { dispatch({type: actions.getLog.reset}) }

export const getLogs = async ({userId, start, limit, currentLogs}) => {
    return async dispatch => {
        const endpoint = `/api/getLogs?id=${userId}${start ? `&skip=${start}` : ''}&limit=${limit}`;

        dispatch({ type: actions.getLogs.request });

        try {
            const response = await axios.get(endpoint);
            const { data, error } = response.data;

            if (error) {
                dispatch({
                    type: actions.getLogs.receive,
                    payload: {
                        isFetching: false,
                        error
                    }
                });

                return;
            }
            
            dispatch({
                type: actions.getLogs.receive,
                payload: {
                    isFetching: false,
                    error: false,
                    logs: currentLogs ? [...currentLogs, ...data] : data,
                    noLogsLeft: !data.length ? true : false
                }
            });
        } catch (error) {
            dispatch({
                type: actions.getLogs.receive,
                payload: {
                    isFetching: false,
                    error
                }
            });
        }
    }
}

export const clearLogs = () => dispatch => { dispatch({ type: actions.getLogs.reset }) }

export const postLog = async ({
    userId,
    date,
    timing,
    rating,
    text
}) => {
    const body = {
        userId,
        date,
        timing,
        rating,
        text
    }

    return async dispatch => {
        try {
            dispatch( {type: actions.postLog.request});
    
            const { data } = await axios.post('/api/postLog', body);
            const { error , success } = data;

            if (error || !success) {
                dispatch({
                    type: actions.postLog.receive,
                    payload: {
                        isPosting : false,
                        success : false,
                        error : error || 'Unknown error occured'
                    }
                });

                return;
            }

            dispatch({
                type: actions.postLog.receive,
                payload: {
                    isPosting : false,
                    success : true,
                    error : null
                }
            })
        } catch (error) {
            dispatch({
                type: actions.postLog.receive,
                payload: {
                    isPosting : false,
                    success : false,
                    error
                }
            })
        }
    }  
}

export const clearLogPost = () => dispatch => { dispatch( {type: actions.postLog.reset})};