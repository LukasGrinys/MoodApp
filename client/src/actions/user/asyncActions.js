import { actions } from './actions';
import axios from 'axios';
import { clearLogs, clearLastLogs } from '../logs/asyncActions';
import Cookie from 'universal-cookie';
const cookie = new Cookie();

export const createUser = async user => {
    return async dispatch => {
        dispatch({type: actions.createUser.request});

        try {
            const { data } = await axios.post('/api/users', user)
            const { success, error } = data;

            if (error || !success) {
                dispatch({
                    type: actions.createUser.receive,
                    payload: {
                        success: false,
                        error: error || 'Could not create user'
                    }
                });

                return;
            }
            
            if (success) {
                dispatch({
                    type: actions.createUser.receive,
                    payload: {
                        success: true,
                        error: null
                    }
                });
            }
        } catch (error) {
            dispatch({
                type: actions.createUser.receive,
                payload: {
                    success: false,
                    error: error || 'Could not create user'
                }
            });
        }
    }
}

export const clearCreateUser = () => dispatch => dispatch({type: actions.createUser.reset});

export const loginUser = async payload => {
    return async dispatch => {
        dispatch({
            type: actions.loginUser.request
        });

        try {
            const { data } = await axios.post('/api/login', payload);
            const { isAuth, error } = data;

            if (error) {
                dispatch({
                    type: actions.loginUser.receive,
                    payload: {
                        isAuth: false,
                        authError : error
                    }
                });

                return;
            }

            if (isAuth) {
                dispatch({
                    type: actions.loginUser.receive,
                    payload: {
                        isAuth,
                        error: null
                    }
                });
            }
        } catch (error) {
            dispatch({
                type: actions.loginUser.receive,
                payload: {
                    isAuth: false,
                    error
                }
            })
        }
    }
}

export function logOut() {
    const request = axios.get('/api/logout');

    return dispatch => {
        request.then( (response) => {
            if (response.status === 200) {
                cookie.remove('auth');
                
                dispatch({
                    type: actions.clearUserData
                });
                
                clearLogs();
                clearLastLogs();
            }
        })
    }
}

export function auth() {
    const response = axios.get('/api/auth')
    .then( ({data}) => data);

    return {
        type: actions.userAuth,
        payload: response
    }
}

export const editUserDetails = async ({userId, details}) => {
    const endpoint = '/api/editAccount';
    const body = {
        id : userId,
        ...details
    }

    return async dispatch => {
        dispatch({type: actions.editUser.request});

        try {
            const { data } = await axios.post(endpoint, body);
            const { error } = data;

            if (error) {
                dispatch({
                    type: actions.editUser.receive,
                    payload: {
                        isEditing: false,
                        success: false,
                        error: error.toString()
                    }
                });

                return
            }

            dispatch({
                type: actions.editUser.receive,
                payload: {
                    isEditing: false,
                    success: true,
                    error: null
                }
            });
        } catch (error) {
            dispatch({
                type: actions.editUser.receive,
                payload: {
                    isEditing: false,
                    success: false,
                    error: error.toString()
                }
            });
        }
    }
}

export const clearEditUser = () => dispatch => {dispatch({type: actions.editUser.reset})}

export const changePassword = async ({
    userId,
    currentPassword,
    newPassword
}) => {
    return async dispatch => {
        if (!userId || !currentPassword || !newPassword) {
            return;
        }

        const endpoint = '/api/changePassword';
        const body = {
            id : userId,
            oldPassword : currentPassword,
            newPassword
        }

        await dispatch({type: actions.changePassword.request});

        try {
            const { error } = await axios.post(endpoint, body);

            if (error) {
                dispatch({
                    type: actions.changePassword.receive,
                    payload: {
                        isChanging : false,
                        error : error,
                        success : false
                    }
                });

                return;
            }

            dispatch({
                type: actions.changePassword.receive,
                payload: {
                    isChanging : false,
                    error : null,
                    success : true
                }
            });
        } catch (error) {
            dispatch({
                type: actions.changePassword.receive,
                payload: {
                    isChanging : false,
                    error : error,
                    success : false
                }
            });
        }
    }
}

export const clearChangePassword = async () => async dispatch => {dispatch({type: actions.changePassword.reset})}

export const clearUserData = async () => async dispatch => {dispatch({type: actions.clearUserData})}