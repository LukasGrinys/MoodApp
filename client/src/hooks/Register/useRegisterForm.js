import { useCallback, useState, useEffect } from 'react';
import { initialFormValues } from '../../constants/initialFormValues';
import * as yup from 'yup';
import { formErrorMessages } from '../../constants/formErrorMessages';
import { useFormik } from 'formik';
import { createUser, loginUser, clearCreateUser } from '../../actions/user/asyncActions';
import { useDispatch, useSelector } from 'react-redux';
import { routerRoutes } from '../../constants/routerRoutes';
import { useHistory } from 'react-router-dom';

export const useRegisterForm = () => {
    const [isSigningUp, setIsSigningUp] = useState(false);
    const dispatch = useDispatch();
    const { createUserError, createUserSuccess } = useSelector( ({user}) => {
        if (user && user.createUser) {
            return {
                createUserError : user.createUser.error,
                createUserSuccess : user.createUser.success
            }
        }

        return {}
    });
    const history = useHistory();

    const handleSignup = useCallback( async values => {
        const payload = {
            firstName: values.firstName,
            lastName: values.lastName,
            email: values.email,
            password: values.password
        };

        setIsSigningUp(true);
        await dispatch(createUser(payload));
        setIsSigningUp(false);
    }, [dispatch]);

    const initialValues = {
        firstName: initialFormValues.firstName,
        lastName: initialFormValues.lastName,
        email: initialFormValues.email,
        password: initialFormValues.password,
        confirmPassword: initialFormValues.confirmPassword
    }

    const validationSchema = yup.object({
        firstName: yup.string().required(formErrorMessages.required),
        lastName: yup.string().required(formErrorMessages.required),
        email: yup
            .string()
            .required(formErrorMessages.required)
            .email(formErrorMessages.email),
        password: yup
            .string()
            .required(formErrorMessages.required)
            .min(6, `${formErrorMessages.min} 6`),
        confirmPassword: yup
            .string()
            .required(formErrorMessages.required)
            .oneOf([yup.ref('password'), null], formErrorMessages.passwordMatch)
    });

    const {
        values,
        errors,
        isValid,
        handleChange,
        handleBlur,
        handleSubmit,
        touched
    } = useFormik({
        initialValues,
        validationSchema,
        onSubmit: handleSignup
    });

    useEffect( () => {
        if (createUserSuccess) {
            const { email, password } = values;

            dispatch(loginUser({
                email,
                password
            })).then( () => {
                history.push(routerRoutes.home);
            });
        }
        // eslint-disable-next-line
    }, [values, createUserSuccess]);

    useEffect( () => {
        return () => {
            dispatch(clearCreateUser());
        }
        // eslint-disable-next-line
    }, [dispatch])

    const controlEvents = {
        handleChange,
        handleBlur
    }

    return {
        values,
        controlEvents,
        errors,
        isValid,
        touched,
        isSigningUp,
        handleSubmit,
        createUserError
    }
}