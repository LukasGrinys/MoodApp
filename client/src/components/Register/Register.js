import React, { Fragment } from 'react';
import { Heading, Wrapper, Label, Input, Footer, ErrorMessage } from '../Forms/Forms';
import Button from '../Button';
import BackButton from '../Back';
import ErrorBox from '../ErrorBox';
import { useRegisterForm } from '../../hooks/Register/useRegisterForm';

const Register = () => {
    const {
        values,
        controlEvents,
        errors,
        isValid,
        touched,
        isSigningUp,
        handleSubmit,
        createUserError
    } = useRegisterForm();

    return (
        <Fragment>
            <Wrapper>   
                <Heading>Sign Up</Heading>
                <form>
                    <Label>Enter your first name</Label>
                    <Input 
                        type="text" 
                        name="firstName"
                        value={values.firstName}
                        controlEvents={controlEvents}
                    />
                    {errors.firstName && touched.firstName && (
                        <ErrorMessage>
                            {errors.firstName}
                        </ErrorMessage>
                    )}

                    <Label>Enter your last name</Label>
                    <Input 
                        type="text" 
                        name="lastName"
                        value={values.lastName}
                        controlEvents={controlEvents}
                    />
                    {errors.lastName && touched.lastName && (
                        <ErrorMessage>
                            {errors.lastName}
                        </ErrorMessage>
                    )}
                    
                    <Label>Enter your email:</Label>
                    <Input 
                        type="email" 
                        name="email"
                        value={values.email}
                        controlEvents={controlEvents}
                    />
                    {errors.email && touched.email && (
                        <ErrorMessage>
                            {errors.email}
                        </ErrorMessage>
                    )}

                    <Label>Enter your password:</Label>
                    <Input 
                        type="password" 
                        name="password"
                        value={values.password}
                        controlEvents={controlEvents}
                    />
                    {errors.password && touched.password && (
                        <ErrorMessage>
                            {errors.password}
                        </ErrorMessage>
                    )}

                    <Label>Confirm your password:</Label>
                    <Input 
                        type="password" 
                        name="confirmPassword"
                        value={values.confirmPassword}
                        controlEvents={controlEvents}
                    />
                    {errors.confirmPassword && touched.confirmPassword && (
                        <ErrorMessage>
                            {errors.confirmPassword}
                        </ErrorMessage>
                    )}

                    <Footer>
                        <Button 
                            color={"primary"}
                            disabled={!isValid || isSigningUp}
                            handleClick={handleSubmit}
                        >
                            Sign up
                        </Button>
                    </Footer>
                </form>
                {  createUserError && (
                    <ErrorBox>
                        Error: {createUserError}
                    </ErrorBox>
                )}
            </Wrapper>
            <BackButton/>
        </Fragment>
    );
};

export default Register;