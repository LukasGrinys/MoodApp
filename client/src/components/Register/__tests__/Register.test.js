import React from 'react';
import { Provider } from 'react-redux';
import 'babel-polyfill';
import Register from '..';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { createUser, loginUser } from '../../../actions/user/asyncActions';
import ErrorBox from '../../ErrorBox';

const mockDispatch = jest.fn();
const mockStore = configureStore();

jest.mock("react-redux", () => {
    const { Provider, useSelector } = jest.requireActual("react-redux");
  
    return {
        useDispatch: () => mockDispatch,
        useSelector,
        Provider
    };
});

jest.mock('react-router-dom', () => {
    return {
        useHistory: jest.fn()
    }
});

const store = mockStore({
    user: {}
});

describe('Register page UI', () => {
    let wrapper;
    beforeEach( () => {
        wrapper = mount(
            <Provider store={store}>
                <Register/>
            </Provider>
        );
    });

    afterEach( () => {
        wrapper.unmount();
    });

    test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('It should have fields for first and last names', () => {
        expect(wrapper.find('input[name="firstName"]').length).toEqual(1);
        expect(wrapper.find('input[name="lastName"]').length).toEqual(1);
    });

    test('First and last name should have proper props', () => {
        expect(wrapper.find('input[name="firstName"]').props()).toEqual({
            type: 'text',
            name: 'firstName',
            value: '',
            className: expect.any(String),
            onChange: expect.any(Function),
            onBlur: expect.any(Function)
        });
        expect(wrapper.find('input[name="lastName"]').props()).toEqual({
            type: 'text',
            name: 'lastName',
            value: '',
            className: expect.any(String),
            onChange: expect.any(Function),
            onBlur: expect.any(Function)
        });
    });

    test('Form should have email and password fields', () => {
        expect(wrapper.find('input[name="email"]').length).toEqual(1);
        expect(wrapper.find('input[name="password"]').length).toEqual(1);
    });

    test('Email and password fields should have proper props', () => {
        expect(wrapper.find('input[name="email"]').props()).toEqual({
            type: 'email',
            name: 'email',
            value: '',
            className: expect.any(String),
            onChange: expect.any(Function),
            onBlur: expect.any(Function)
        })
        expect(wrapper.find('input[name="password"]').props()).toEqual({
            type: 'password',
            name: 'password',
            value: '',
            className: expect.any(String),
            onChange: expect.any(Function),
            onBlur: expect.any(Function)
        })
    });

    test('It should have a Sign Up button', () => {
        expect(wrapper.find('button').length).toEqual(1);
    });

    test('Sign up button should have proper props', () => {
        expect(wrapper.find('button').props()).toEqual({
            className: expect.any(String),
            onClick: expect.any(Function),
            disabled: false,
            type: 'button',
            children: 'Sign up'
        })
    });
});

describe('Register form logic', () => {
    const firstName = 'Test';
    const lastName = 'Testington';
    const email = 'mail@mail.com';
    const password = 'password';

    let wrapper;
    beforeEach( () => {
        wrapper = mount(
            <Provider store={store}>
                <Register/>
            </Provider>
        );
    });

    afterEach( () => {
        wrapper.unmount();
    });

    test('Initial form values are empty', () => {
        expect(wrapper.find('input[name="firstName"]').props().value).toEqual('');
        expect(wrapper.find('input[name="lastName"]').props().value).toEqual('');
        expect(wrapper.find('input[name="email"]').props().value).toEqual('');
        expect(wrapper.find('input[name="password"]').props().value).toEqual('');
    });

    test('Input values change when typed', () => {
        wrapper.find('input[name="firstName"]').simulate('change', {
            target : {
                name: 'firstName',
                value: firstName
            }
        });
        wrapper.find('input[name="lastName"]').simulate('change', {
            target : {
                name: 'lastName',
                value: lastName
            }
        });

        wrapper.update();

        expect(wrapper.find('input[name="firstName"]').prop('value')).toEqual(firstName);
        expect(wrapper.find('input[name="lastName"]').prop('value')).toEqual(lastName);
    });

    test('Button should be disabled when values are incorrect', () => {
        wrapper.find('input[name="firstName"]').simulate('blur');

        setTimeout(() => {
            wrapper.update();
            const button = wrapper.find('button'); 
            expect(button.prop('disabled')).toBeTruthy();
            done();
        }, 0);
    });

    test('Error message should be shown if values are incorrect', () => {
        wrapper.find('input[name="firstName"]').simulate('blur');

        setTimeout(() => {
            wrapper.update();
            expect(wrapper.find(/errorMessage/i)).toHaveLengthOf(1);
            done();
        }, 0);
    });

    test('Button should be enabled when values are correct', () => {
        wrapper.find('input[name="firstName"]').simulate('change', {
            target : {
                name: 'firstName',
                value: firstName
            }
        });
        wrapper.find('input[name="lastName"]').simulate('change', {
            target : {
                name: 'lastName',
                value: lastName
            }
        });
        wrapper.find('input[name="email"]').simulate('change', {
            target : {
                name: 'email',
                value: email
            }
        });
        wrapper.find('input[name="lastName"]').simulate('change', {
            target : {
                name: 'password',
                value: password
            }
        });

        setTimeout(() => {
            wrapper.update();
            const button = wrapper.find('button'); 
            expect(button.prop('disabled')).toBeFalsy();
            done();
        }, 0);
    });

    test('Register action is called when submitting form with proper values', () => {
        wrapper.find('input[name="firstName"]').simulate('change', {
            target : {
                name: 'firstName',
                value: firstName
            }
        });
        wrapper.find('input[name="lastName"]').simulate('change', {
            target : {
                name: 'lastName',
                value: lastName
            }
        });
        wrapper.find('input[name="email"]').simulate('change', {
            target : {
                name: 'email',
                value: email
            }
        });
        wrapper.find('input[name="lastName"]').simulate('change', {
            target : {
                name: 'password',
                value: password
            }
        });
        wrapper.find('button').simulate('click'); 

        setTimeout( () => {
            wrapper.update();
            expect(createUser).toHaveBeenCalledTimes(1);
            done();
        }, 0)
    })
});

describe('Register form error/success handling', () => {
    let wrapper;

    test('Shows error message if registering was not successful', () => {
        const testError = 'Test error';

        const mockedStore = mockStore({
            user: {
                createUser: {
                    success : false,
                    error : testError
                }
            }
        })

        const wrapper = mount(
            <Provider store={mockedStore}>
                <Register/>
            </Provider>
        );

        expect(wrapper.find(ErrorBox).length).toBe(1);
        expect(wrapper.find(ErrorBox).text()).toEqual(`Error: ${testError}`);
    });

    test('User is logged in if registration was succesful', () => {
        const mockedStore = mockStore({
            user: {
                createUser: {
                    success : true,
                    error : false
                }
            }
        })

        mount(
            <Provider store={mockedStore}>
                <Register/>
            </Provider>
        );

        setTimeout( () => {
            expect(loginUser).toHaveBeenCalledTimes(1);
            done();
        }, 0)
    });

})
