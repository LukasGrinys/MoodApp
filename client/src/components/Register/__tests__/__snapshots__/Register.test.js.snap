// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register page UI renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Register>
    <Wrapper>
      <div
        className="formWrapper"
      >
        <Heading>
          <h1
            className="heading"
          >
            Sign Up
          </h1>
        </Heading>
        <form>
          <Label>
            <span
              className="label"
            >
              Enter your first name
            </span>
          </Label>
          <Input
            controlEvents={
              Object {
                "handleBlur": [Function],
                "handleChange": [Function],
              }
            }
            name="firstName"
            placeholder=""
            type="text"
            value=""
          >
            <input
              className="textInput"
              name="firstName"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </Input>
          <Label>
            <span
              className="label"
            >
              Enter your last name
            </span>
          </Label>
          <Input
            controlEvents={
              Object {
                "handleBlur": [Function],
                "handleChange": [Function],
              }
            }
            name="lastName"
            placeholder=""
            type="text"
            value=""
          >
            <input
              className="textInput"
              name="lastName"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </Input>
          <Label>
            <span
              className="label"
            >
              Enter your email:
            </span>
          </Label>
          <Input
            controlEvents={
              Object {
                "handleBlur": [Function],
                "handleChange": [Function],
              }
            }
            name="email"
            placeholder=""
            type="email"
            value=""
          >
            <input
              className="textInput"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              type="email"
              value=""
            />
          </Input>
          <Label>
            <span
              className="label"
            >
              Enter your password:
            </span>
          </Label>
          <Input
            controlEvents={
              Object {
                "handleBlur": [Function],
                "handleChange": [Function],
              }
            }
            name="password"
            placeholder=""
            type="password"
            value=""
          >
            <input
              className="textInput"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              type="password"
              value=""
            />
          </Input>
          <Label>
            <span
              className="label"
            >
              Confirm your password:
            </span>
          </Label>
          <Input
            controlEvents={
              Object {
                "handleBlur": [Function],
                "handleChange": [Function],
              }
            }
            name="confirmPassword"
            placeholder=""
            type="password"
            value=""
          >
            <input
              className="textInput"
              name="confirmPassword"
              onBlur={[Function]}
              onChange={[Function]}
              type="password"
              value=""
            />
          </Input>
          <Footer>
            <div
              className="formFooter"
            >
              <Button
                color="primary"
                disabled={false}
                handleClick={[Function]}
              >
                <button
                  className="button btnPrimary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Sign up
                </button>
              </Button>
            </div>
          </Footer>
        </form>
      </div>
    </Wrapper>
    <BackButton>
      <div
        className="container"
      >
        <div
          className="button"
          onClick={[Function]}
        >
          <FontAwesome
            name="arrow-left"
          >
            <span
              aria-hidden={true}
              className="fa fa-arrow-left"
            />
          </FontAwesome>
          <div
            className="text"
          >
            Go back
          </div>
        </div>
      </div>
    </BackButton>
  </Register>
</Provider>
`;
